projects/libs/core-library-typescript/src/consensus/block-builder.ts

```diff
@@ -1,36 +1,103 @@
-import { types } from "../common-library";
+import { types, logger, BlockUtils } from "../common-library";
 
 export default class BlockBuilder {
-    private virtualMachine: types.VirtualMachineClient;
-    private transactionPool: types.TransactionPoolClient;
+  private virtualMachine: types.VirtualMachineClient;
+  private transactionPool: types.TransactionPoolClient;
+  private pollIntervalMs: number;
+  private pollInterval: NodeJS.Timer;
+  private lastBlock: types.Block;
+  private readyForBlockAppend = false;
+  private blockStorage: types.BlockStorageClient;
+  private onNewBlockBuild: (block: types.Block) => void;
 
-    constructor(input: {virtualMachine: types.VirtualMachineClient, transactionPool: types.TransactionPoolClient}) {
-        this.virtualMachine = input.virtualMachine;
-        this.transactionPool = input.transactionPool;
+  constructor(input: {
+    virtualMachine: types.VirtualMachineClient,
+    transactionPool: types.TransactionPoolClient,
+    blockStorage: types.BlockStorageClient,
+    newBlockBuildCallback: (block: types.Block) => void,
+    pollIntervalMs?: number
+  }) {
+      this.virtualMachine = input.virtualMachine;
+      this.transactionPool = input.transactionPool;
+      this.blockStorage = input.blockStorage;
+      this.onNewBlockBuild = input.newBlockBuildCallback;
+      this.pollIntervalMs = input.pollIntervalMs || 500;
+  }
+
+  private pollForPendingTransactions() {
+    this.pollInterval = setInterval(async () => {
+      try {
+        if (this.readyForBlockAppend) {
+          logger.debug("blockBuilder tick");
+          await this.appendNextBlock();
+        }
+      } catch (err) {
+        logger.error("newBlockAppendTick error: " + err);
+      }
+    }, this.pollIntervalMs);
+  }
+
+  private async buildBlockFromPendingTransactions(lastBlock: types.Block): Promise<types.Block> {
+    const { transactions } = await this.transactionPool.getAllPendingTransactions({});
+
+    if (transactions.length == 0) {
+        throw "transaction pool is empty";
     }
 
-    public async buildNextBlock(lastBlockId: number) {
-        const { transactions } = await this.transactionPool.getAllPendingTransactions({});
+    const { processedTransactions, stateDiff } = await this.virtualMachine.processTransactionSet({ orderedTransactions: transactions });
 
-        if (transactions.length == 0) {
-            throw new Error("Transaction pool is empty");
-        }
+    if (processedTransactions.length == 0) {
+        throw new Error("Transaction pool is empty");
```

Comment:
@netoneko discovered an issue with this approach: unprocessed transactions (regardless if all or only some) aren't removed from the transaction pool(s). Could you address it as well?